source/slang/slang-parser.cpp

```diff
@@ -4310,6 +4361,130 @@ namespace Slang
+                // so this *only* works if the type is defined before the cast.
+                
+                return (_tryResolveTypeDecl(parser, expr) != nullptr);
+            }
+            case TokenType::OpAdd:
+            case TokenType::OpSub:
+            {
+                // + - are ambiguous, it could be a binary + or - so -> expression, or unary -> cast
+                //
+                // (Some::Stuff) + 3
+                // (Some::Stuff) - 3
+                // Strictly I can only tell if this is an expression or a cast if I know Some::Stuff is a type or not
+                // but we can't know here in general because we allow out-of-order declarations.
+
+                // If we can determine it's a type, then it must be a cast, and we are done.
+                // 
+                // NOTE! This test can only determine if it's a type *iff* it has already been defined. A future out
+                // of order declaration, will not be correctly found here.
+                //
+                // This means the semantics change depending on the order of definition (!) 
+                if (_tryResolveTypeDecl(parser, expr))
+                {
+                    return true;
+                }
+
+                // Now we use a heuristic. If there is no white space between the + or - and what follows
+                // we will assume they are unary ops, and therefore this is a cast.
+                // Otherwise we assume they are binary ops.
+                //
+                // Ie:
+                // (Some::Stuff) +3  - must be a cast
+                // (Some::Stuff) + 3 - must be an expression.
+
+                // TODO(JS): This covers the (SomeScope::Identifier) case
+                //
+                // But perhaps there other ways of referring to types, that this now misses? With associated types/generics perhaps.
+                // 
+                // For now we'll assume it's not a cast if it's not a StaticMemberExpr
+                // The reason for the restriction (which perhaps can be broadened), is we don't
+                // want the interpretation of something in parentheses to be determined by something as common as + or - whitespace.
+
+                if (auto staticMemberExpr = dynamicCast<StaticMemberExpr>(expr))
+                {
+                    // Apply the heuristic
+
+                    TokenReader::ParsingCursor cursor = parser->tokenReader.getCursor();
+                    // Skip the + or -
+                    advanceToken(parser);
+                    // Peek the next token to see if it was preceded by white space
+                    const Token nextToken = peekToken(parser);
+
+                    // Rewind
+                    parser->tokenReader.setCursor(cursor);
+
+                    // If there isn't any whitespace prior, we assume unary, and that means it must be a cast
+                    return ((nextToken.flags & TokenFlag::AfterWhitespace) == 0);
```

Comment:
I believe that the correct heuristic for identifying whether an operator is binary or unary should be something like:

* Whitespace before, whitespace after -> binary
* No whitespace before, no whitespace after -> binary
* Otherwise -> unary

This basically breaks things down as:

```hlsl
a + b // assume binary
a+b // assume binary
a+ b // assume unary (looks like a postfix op)
a +b // assume unary (looks like a prefix op)
```

Reply:
Changed to this style.